client.go

```diff
@@ -519,109 +513,118 @@ func WithImportLabels(labels map[string]string) ImportOpt {
 	}
 }
 
-// WithOCIImportFormat sets the import format for an OCI image format
-func WithOCIImportFormat() ImportOpt {
+// WithImporter sets the importer
+func WithImporter(importer imageformats.Importer) ImportOpt {
 	return func(c *importOpts) error {
-		if c.format != "" {
-			return errors.New("format already set")
+		if c.importer != nil {
+			return errors.New("impoter already set")
 		}
-		c.format = ociImageFormat
-		return nil
-	}
-}
-
-// WithRefObject specifies the ref object to import.
-// If refObject is empty, it is copied from the ref argument of Import().
-func WithRefObject(refObject string) ImportOpt {
-	return func(c *importOpts) error {
-		c.refObject = refObject
+		c.importer = importer
 		return nil
 	}
 }
 
-func resolveImportOpt(ref string, opts ...ImportOpt) (importOpts, error) {
+func resolveImportOpt(opts ...ImportOpt) (importOpts, error) {
 	var iopts importOpts
 	for _, o := range opts {
 		if err := o(&iopts); err != nil {
 			return iopts, err
 		}
 	}
 	// use OCI as the default format
-	if iopts.format == "" {
-		iopts.format = ociImageFormat
-	}
-	// if refObject is not explicitly specified, use the one specified in ref
-	if iopts.refObject == "" {
-		refSpec, err := reference.Parse(ref)
-		if err != nil {
-			return iopts, err
-		}
-		iopts.refObject = refSpec.Object
+	if iopts.importer == nil {
+		iopts.importer = oci.V1Importer
 	}
 	return iopts, nil
 }
 
 // Import imports an image from a Tar stream using reader.
 // OCI format is assumed by default.
+// Filter strings are format-specific, but needs to match exactly one item.
+//
+// TODO(AkihiroSuda): support importing multiple images from a single archive at once? (map[ref]filter?)
 //
-// Note that unreferenced blobs are imported to the content store as well.
-func (c *Client) Import(ctx context.Context, ref string, reader io.Reader, opts ...ImportOpt) (Image, error) {
-	iopts, err := resolveImportOpt(ref, opts...)
+// Note that unreferenced blobs may be imported to the content store as well.
+func (c *Client) Import(ctx context.Context, ref string, reader io.Reader, filter string, opts ...ImportOpt) (Image, error) {
```

Comment:
On second thought, should I remove filters, and change this function definition to `Import (ctx context.Context, reader io.Reader, opts ...ImportOpt)(map[string]Image, error)`?

The map key for `map[string]Image` is reference string, which is determined by the importer implementation.
(However, it may override existing refs...)

Reply:
What would the prefix represent? Various importers?

Reply:
@AkihiroSuda What is this filter used for? Selecting an image from the tarball when there are multiple images inside the tar? I didn't quite get it.

Reply:
I'm suggesting remove filters here to support loading multiple images from a single archive at once. https://github.com/containerd/containerd/pull/1602#discussion_r143395705

Reply:
Hm, the prefix part seems confusing to me. :P Are you trying to avoid conflict refs inside the tarball? Or conflict ref with existing ref in containerd?

If it's the later, how does Docker solve it today?

Reply:
Latter and it is not resolved in Docker (just override).

For importing v1 image, you could omit prefix with risk of overriding.

But for importing OCI image, we somehow need to pass extra prefix (and default tag?), because OCI images lack `RepoTags`-equivalent.